src/api.ts

```diff
@@ -0,0 +1,124 @@
+const getRandomGenre = () => {
+  return genres.splice(Math.floor(Math.random()*genres.length), 1)[0];
+};
+
+// creates an image element from a supplied url
+const loadImage = (imageUrl: string): Promise<HTMLImageElement> => {
+  return new Promise((resolve, reject) => {
+    const img = new Image();
+    img.onload = () => resolve(img);
+    img.onerror = () => reject(new Error(`load ${imageUrl} fail`));
+    img.src = imageUrl;
+  });
+};
+
+// generates the rows data by random grabbing lists of popular titles from the list of genres supplied by the movie db
+export const getRowsData = (defaultImage: HTMLImageElement, IMAGE_WIDTH: number, IMAGE_HEIGHT: number, PADDING: number) => {
+  return new Promise((resolve, reject) => {
+    fetchGenres().then((res) => {
+      genres = res.genres;
+      const intialRows = [];
+      const theRowsData: Array<RowObjectType> = [];
+      intialRows.push(getRandomGenre());
+      intialRows.push(getRandomGenre());
+      intialRows.push(getRandomGenre());
+      intialRows.push(getRandomGenre());
+      let rowCounter = 0;
+      intialRows.forEach((rowIn) => {
+        fetchGenreRow(rowIn.id).then((res) => {
+          const { results } = res;
+          results.length = 10; // TODO only load assets that are going to be seen. Temporarily chopping the array in half until this is added
+          const cards = res.results.map((card, cardIndex) => {
+            const { original_title, overview, backdrop_path, poster_path, release_date } = card;
+            return {
+              image: defaultImage,
+              cardOriginalPositionX: cardIndex * (IMAGE_WIDTH + PADDING),
+              cardOriginalPositionY: rowCounter * (IMAGE_HEIGHT + PADDING),
+              original_title,
+              overview,
+              backdrop_path,
+              poster_path,
+              release_date,
+            } as RowImageType;
+          })
+          const row: RowObjectType = {
+            id: rowIn.id,
+            cards,
+            rowNumber: rowCounter,
+            highlightedCard: 0,
+            translateX: 0, // position offset from starting point
+            animationStartTime: null,
+            animationDirection: null,
+            targetTranslateX: 0, // the end translateX position of the animation upon completion of animation
+            easingPosition: 0, // how far through x animation row is. Used for calculating scaling of highlighted image
+            unfinishedMovementX: 0, // if we interrupt an X movement we need to include this in the new animation start position
+          }
+          rowCounter +=1 ;
+          theRowsData.push(row)
+          if (theRowsData.length === intialRows.length) {
+            resolve(theRowsData);
+          }
+        })
+      })
+    });
+  })
+};
+
+
+// loads and re-renders images
+export const loadAndRenderImages = (rowsData: Array<RowObjectType>, ctx: CanvasRenderingContext2D, translateY: number ) => {
```

Comment:
By the name of this function, I think we could file an issue with further refactoring here.

Reply:
Yeah absolutely. I thought a fair bit about this and I don't have a simpler solution yet that doesn't add additional complexity.
As soon as the image is loaded - we need to update the image on the carousel to replace and blank spots.
Missing feature - the function should check if an animation is in progress, if so then rendering is not necessary as the new image will display on the next frame of the animation.